### PR TITLE
Refactor/boas 1189 review document api code

### PR DESCRIPTION
**Pull request by DavidJWhiteDev**
## Describe your changes

Refactoring of the Document and S51 Advice APIs to:
- correct type paths following code restructure eg changing `@typedef {import('apps/api/src/database/schema.js').Document} Document` to `@typedef {import('@prisma/client').Document} Document`
- use new shortcuts for imports rather than log relative paths is poss eg 
-- #repositories/
-- #utils/ 
-- #middleware/
- ensure swagger types correctly match the API payloads - both request and response
- fix errors in OpenAPI endpoint decorations so that they work correctly in the swagger api page /api-docs
- clean up some old-style type definitions in swagger.js, and add some new types as appropriate
- make use of the new swagger-types file to use the swagger type definitions as actual types to be used in code

## BOAS-1189 Refactor Doc APIs
https://pins-ds.atlassian.net/browse/BOAS-1189

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [x] I have made corresponding changes to the documentation
- [x] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [x] I have provided a description of the ticket
- [x] I have included unit tests to cover any testable code changes

### CODE REVIEW

**Review by francescobrenci-AR — APPROVED** [review]

File: apps/api/src/server/applications/application/documents/document.controller.js [hunk 1]
```diff
@@ -30,21 +32,25 @@ import {
...
+ * @typedef {import('@pins/applications.api').Api.DocumentToSave} DocumentToSave
+ * @typedef {import('@pins/applications.api').Api.DocumentToSaveExtended} DocumentToSaveExtended
+ * @typedef {import('@pins/applications.api').Api.DocumentsToSaveManyRequestBody} DocumentsToSaveManyRequestBody
  */
 
 /**
- * @type {import('express').RequestHandler<any, any, { blobStorageHost: string, privateBlobContainer: string, documents: { documentName: string, blobStoreUrl: string }[] } | any, any>}
+ * Upload an array of documents to a folder on a case
+ *
+ * @type {import('express').RequestHandler<any, any, { DocumentsToSaveManyRequestBody } | any, any>}
```
Comment (beejjacobs) [inline apps/api/src/server/applications/application/documents/document.controller.js:49]: I think this is meant to be `DocumentsToSaveManyRequestBody | any` (I'm assuming here the body is meant to be an array).

File: apps/api/src/server/applications/application/documents/document.js [hunk 2]
```diff
@@ -58,8 +58,11 @@ export const buildNsipDocumentPayload = (version) => {
 		originalFilename: version.originalFilename,
 		size: version.size,
 		mime: version.mime,
-		documentURI: buildBlobUri(version.privateBlobContainer, version.privateBlobPath),
-		publishedDocumentURI: buildBlobUri(version.publishedBlobContainer, version.publishedBlobPath),
+		documentURI: buildBlobUri(version.privateBlobContainer ?? '', version.privateBlobPath ?? ''),
```
Comment (beejjacobs) [inline apps/api/src/server/applications/application/documents/document.js:61]: Is it neater to update `buildBlobUri` to accept `string | null`? It seems to already handle it.

File: apps/api/src/server/applications/application/documents/document.service.js [hunk 3]
```diff
@@ -217,80 +225,63 @@ const upsertDocumentVersionsMetadataToDatabase = async (
...
 	await upsertDocumentVersionsMetadataToDatabase(
 		responseFromDocumentStorage.documents,
 		responseFromDocumentStorage.privateBlobContainer
 	);
 
-	// @ts-ignore
-	/**
-	 * @type {any[]}
-	 */
+	/** @type {DocumentAndBlobInfoManyResponse[]} */
```
Comment (beejjacobs) [inline apps/api/src/server/applications/application/documents/document.service.js:280]: This isn't an array of responses, but just an array of promises for all the activity logs.

i.e. `Promise<import('@prisma/client').DocumentActivityLog>[]`

File: apps/api/src/server/applications/application/documents/document.service.js [hunk 4]
```diff
@@ -217,80 +225,63 @@ const upsertDocumentVersionsMetadataToDatabase = async (
...
 	const documentActivityLogs = [];
 	requestToDocumentStorage.forEach((document) => {
```
Comment (beejjacobs) [inline apps/api/src/server/applications/application/documents/document.service.js:282]: Separate ticket, but: wouldn't it make sense to use createMany?

File: apps/api/src/server/applications/application/documents/document.service.js [hunk 5]
```diff
@@ -217,80 +225,63 @@ const upsertDocumentVersionsMetadataToDatabase = async (
...
 		documentActivityLogs.push(
+			// @ts-ignore
```
Comment (beejjacobs) [inline apps/api/src/server/applications/application/documents/document.service.js:284]: Won't need this if we fix type as above.

File: apps/api/src/server/utils/document-storage.js [hunk 6]
```diff
@@ -18,6 +18,7 @@ import config from '../config/config.js';
  * @typedef {DocumentInfo & WithBlobUrl} DocumentInfoWithBlobUrl
  */
 
+// TODO: DJW should we not use  * @returns {Promise<import('@pins/applications.api').Api.DocumentAndBlobInfoManyResponse>}
```
Comment (beejjacobs) [inline apps/api/src/server/utils/document-storage.js:21]: Yes, makes sense to me! When I refactored this I tried to make as few changes as possible. Let's use that type, and remove the above type defs that we shouldn't need.

File: docs/types.md [hunk 7]
```diff
@@ -23,5 +23,5 @@ import('@prisma/client').Prisma.MyTypeUncheckedUpdateInput
 If you need a type with a relation included, say a `ProjectUpdate` with a `Case`, you can use:
 
 ```
-import('@prisma/client').Prisma.ProjectUpdateGetPayload<{include: {case: true}}
+import('@prisma/client').Prisma.ProjectUpdateGetPayload<{include: {case: true}}>
```
Comment (beejjacobs) [inline docs/types.md:26]: Good spot!

**Review by beejjacobs — CHANGES_REQUESTED** [review]
@DavidJWhiteDev this is excellent, very thorough and I am sure took a good amount of time.

I've got a few comments/suggestions - happy to discuss.

I like the use of the subpath imports, and importing the types needed at the top of a file. Makes things easier to follow for sure, and good to have correct types. I also like the naming of the API request/response types which clearer describe what they include/are for. Good work!

File: apps/api/src/server/applications/application/documents/document.service.js (hunk 4 above)
Comment (DavidJWhiteDev) [reply apps/api/src/server/applications/application/documents/document.service.js:282]: TODO: comment added for future

**Review by beejjacobs — APPROVED** [review]
All issues resolved, LGTM!

**Pull request merged**